mirrord-layer/src/sockets.rs

```diff
@@ -1,240 +1,506 @@
+unsafe extern "C" fn socket_detour(domain: c_int, type_: c_int, protocol: c_int) -> c_int {
+    socket(domain, type_, protocol)
+}
 
-        if let Some(mut socket) = sockets.remove(&sockfd) {
-            let mut buffer = [0; 1];
-            socket.read_socket.read_exact(&mut buffer).unwrap();
-            sockets.insert(sockfd, socket.address.port(), socket);
-            let mut queues = self.connection_queues.lock().unwrap();
-            let queue = queues.get_mut(&sockfd).unwrap();
-            queue.remove().unwrap()
-        } else {
-            error!("No socket found for fd: {}", sockfd);
-            0
+/// Check if the socket is managed by us, if it's managed by us and it's not an ignored port,
+/// update the socket state and don't call bind (will be called later). In any other case, we call
+/// regular bind.
+fn bind(sockfd: c_int, addr: *const sockaddr, addrlen: socklen_t) -> c_int {
+    debug!("bind called sockfd: {:?}", sockfd);
+    let mut socket = {
+        let mut sockets = SOCKETS.lock().unwrap();
+        match sockets.take(&sockfd) {
+            Some(socket) if !matches!(socket.state, SocketState::Initialized) => {
+                error!("socket is in invalid state for bind {:?}", socket.state);
+                return libc::EINVAL;
+            }
+            Some(socket) => socket,
+            None => {
+                debug!("bind: no socket found for fd: {}", &sockfd);
+                return unsafe { libc::bind(sockfd, addr, addrlen) };
+            }
+        }
+    };
+
+    let raw_addr = unsafe { OsSocketAddr::from_raw_parts(addr as *const u8, addrlen as usize) };
+    let parsed_addr = match raw_addr.into_addr() {
+        Some(addr) => addr,
+        None => {
+            error!("bind: failed to parse addr");
+            return libc::EINVAL;
         }
+    };
+
+    debug!("bind:port: {}", parsed_addr.port());
+    if is_ignored_port(parsed_addr.port()) {
+        debug!("bind: ignoring port: {}", parsed_addr.port());
+        return unsafe { libc::bind(sockfd, addr, addrlen) };
     }
 
-    pub fn open_connection(&self, connection_id: ConnectionId, port: Port) {
-        let mut connections = self.connections.lock().unwrap();
-        if let Some(mut socket) = connections.remove_alt(&port) {
-            debug!("new connection id: {:?}", connection_id);
-            let mut queues = self.connection_queues.lock().unwrap();
-            match queues.get_mut(&socket.read_fd) {
-                Some(queue) => {
-                    queue.add(connection_id).unwrap();
+    socket.state = SocketState::Bound(Bound {
+        address: parsed_addr,
+    });
+
+    let mut sockets = SOCKETS.lock().unwrap();
+    sockets.insert(socket);
+    0
+}
+
+unsafe extern "C" fn bind_detour(
+    sockfd: c_int,
+    addr: *const sockaddr,
+    addrlen: socklen_t,
+) -> c_int {
+    bind(sockfd, addr, addrlen)
+}
+
+fn listen(sockfd: RawFd, _backlog: c_int) -> c_int {
```

Comment:
Bind the socket to a fake, local port, and subscribe to the agent on the real port. Messages received from the agent on the real port will later be routed to the fake local port.